openhands_cli/acp_impl/agent.py

```diff
@@ -0,0 +1,477 @@
+            # Callbacks are already set up when conversation was created
+            # Track the running task so cancel() can wait for proper cleanup
+            run_task = asyncio.create_task(asyncio.to_thread(conversation.run))
+            self._running_tasks[session_id] = run_task
+            try:
+                await run_task
+            finally:
+                # Clean up task tracking
+                self._running_tasks.pop(session_id, None)
+
+            # Return the final response
+            return PromptResponse(stopReason="end_turn")
+
+        except RequestError:
+            # Re-raise RequestError as-is
+            raise
+        except Exception as e:
+            logger.error(f"Error processing prompt: {e}", exc_info=True)
+            # Send error notification to client
+            await self._conn.sessionUpdate(
+                SessionNotification(
+                    sessionId=session_id,
+                    update=AgentMessageChunk(
+                        sessionUpdate="agent_message_chunk",
+                        content=TextContentBlock(type="text", text=f"Error: {str(e)}"),
+                    ),
+                )
+            )
+            raise RequestError.internal_error(
+                {"reason": "Failed to process prompt", "details": str(e)}
+            )
+
+    async def cancel(self, params: CancelNotification) -> None:
+        """Cancel the current operation."""
+        logger.info(f"Cancel requested for session: {params.sessionId}")
+
+        try:
+            # Get active conversation and pause it
+            conversation = self._get_or_create_conversation(session_id=params.sessionId)
+            # Pause the conversation (state is preserved in cache)
+            conversation.pause()
+
+            # Wait for the running task to actually terminate
+            # This ensures we're faithful about the conversation status
+            running_task = self._running_tasks.get(params.sessionId)
+            if running_task and not running_task.done():
+                logger.debug(
+                    f"Waiting for conversation thread to terminate for session "
+                    f"{params.sessionId}"
+                )
+                try:
+                    # Wait for the task to complete with a timeout
+                    # The pause() signal should cause it to stop soon
+                    await asyncio.wait_for(running_task, timeout=10.0)
+                except TimeoutError:
+                    logger.warning(
+                        f"Conversation thread did not stop within timeout for session "
+                        f"{params.sessionId}"
+                    )
+                    # Cancel the task if it doesn't stop in time
+                    running_task.cancel()
+                    try:
+                        await running_task
+                    except asyncio.CancelledError:
+                        pass
+                except Exception as e:
```

Comment:
NIT: should this raise a ACP `RequestError`?

Reply:
@OpenHands please investigate, if so, help me modify code.